### PR TITLE
Organise steps logic

### PR DESCRIPTION
**Pull request by tcapelle**
The idea is showing:
- A progress bar with the actual total count
- Having the same steps logged and reported on the progress bar
- Count a training step as an optimizer step to be consistent with other libraries like transformers.Trainer and Axolotl.
- Add some QPS and grad norm

### CODE REVIEW

**Comment by tcapelle** [comment]
Are these they types of QPS you were expecting @msaroufim ?

File: recipes/full_finetune_single_device.py [hunk 1]
```diff
@@ -409,46 +422,58 @@ def train(self) -> None:
...
                     self._optimizer.step()
                     self._optimizer.zero_grad(set_to_none=True)
 
+                    pbar.update(1)
+                    pbar.set_description(
+                        f"{curr_epoch+1}|{self.total_training_steps+1}|Loss: {loss.item()}"
+                    )
+
+                    # compute training metrics
+                    if self.total_training_steps % self._log_every_n_steps == 0:
```
[the remark this answers is not in the document]
Comment (tcapelle) [reply recipes/full_finetune_single_device.py:442]: I think is what should be no? we log on actual optimizers steps

**Comment by tcapelle** [comment]
Running this command:
```bash
tune run full_finetune_single_device --config my_dev/tiny_llama.yaml log_every_n_steps=5 max_steps_per_epoch=10 epochs=2
```
you get a progress bar with only 10 steps:
```
INFO:torchtune.utils.logging:Dataset and Sampler are initialized.
1|10|Loss: 10.452272415161133: 100%|█████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  3.65it/s]INFO:torchtune.utils.logging:Model checkpoint of size 0.08 GB saved to my_dev/tiny_llama/checkpoints/torchtune_model_0.pt
INFO:torchtune.utils.logging:Recipe checkpoint of size 0.16 GB saved to my_dev/tiny_llama/checkpoints/recipe_state.pt
1|10|Loss: 10.452272415161133: 100%|█████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  2.85it/s]
  0%|                                                                                                         | 0/10 [00:00<?, ?it/sINFO:torchtune.utils.logging:Model checkpoint of size 0.08 GB saved to my_dev/tiny_llama/checkpoints/torchtune_model_1.pt0,  1.72it/s]
2|20|Loss: 10.449804306030273: 100%|█████████████████████████████████████████████████████████████████| 10/10 [00:05<00:00,  1.78it/s]
```

and you get only 2 logs per epoch:
```
Step 0 | loss:10.565682411193848 lr:2e-05 gpu_resources:0 tokens_per_second:456.04617469323597 iterations_per_second:1.6523412126566521 grad_norm:16.26054121046421 
Step 5 | loss:10.421597480773926 lr:2e-05 gpu_resources:0 tokens_per_second:620.2041065388681 iterations_per_second:3.828420410733754 grad_norm:23.697011790378774 
Step 10 | loss:10.490856170654297 lr:2e-05 gpu_resources:0 tokens_per_second:376.36752451194593 iterations_per_second:1.344169730399807 grad_norm:15.191259666319432 
Step 15 | loss:10.376498222351074 lr:2e-05 gpu_resources:0 tokens_per_second:1275.976294664742 iterations_per_second:3.2385185143775175 grad_norm:12.678521182631085 
```

**Comment by RdoubleA** [comment]
Confirmed that this works as expected with optimizer_in_bwd=True and gradient accumulation off, and optimizer_in_bwd=False and gradient accumulation on:

`tune run full_finetune_single_device --config llama2/7B_full_low_memory optimizer_in_bwd=False gradient_accumulation_steps=5 max_steps_per_epoch=10`
`tune run full_finetune_single_device --config llama2/7B_full_low_memory`

**Comment by tcapelle** [comment]
I added the changes you put on running and computing the right amount of tokens. I am not sure the GPU metrics is something you want to log very often as it is expensive to compute. Do you know wy the grad accum test doesn't pass?

**Comment by ebsmothers** [comment]
> I am not sure the GPU metrics is something you want to log very often as it is expensive to compute. 

We could consider adding a flag like `log_memory_stats` to the config to enable/disable this. What do you think?

> Do you know wy the grad accum test doesn't pass?

It has to do with how we infer the loss for the test based on the logs (see [here](https://github.com/pytorch/torchtune/blob/52567c9ec085f60c7f4e5bb8dcacb0bc1ce1bec7/tests/recipes/test_full_finetune_single_device.py#L215-L239)). Previously, whenever `self.total_training_steps % self._log_every_n_steps == 0`, we would log on every iteration of that step, so we would actually get a list of loss values (hence the np.mean in L239 of the above code pointer). With your change we only logged on the final iteration of the step, so the logged value would just be the loss on the last iteration, not the average over all the iterations. This is why I added running loss instead, to accumulate losses over each iteration and then log the mean for the step.

**Comment by RdoubleA** [comment]
closing this as it was addressed in #831
